python/lambda-ddb-mysql-etl-pipeline/lambda/asteroids.py

```diff
@@ -0,0 +1,155 @@
+        # method call below used when Secrets Manager integrated. See get_secret.__doc__ for more.
+        # self.api_key=get_secret('nasa_api_key')
+        self.api_key=os.environ["NASA_KEY"]
+        self.endpoint=f"""https://api.nasa.gov/neo/rest/v1/feed?start_date={self.today}&end_date={self.today}&api_key={self.api_key}"""
+        self.response_object=self.nasa_client(self.endpoint)
+        self.processed_response=self.process_asteroids(self.response_object)
+        if self.file_format == "json":
+            self.asteroids_json(self.processed_response)
+        elif self.file_format == "csv":
+            self.asteroids_csv(self.processed_response)
+        else:
+            raise Exception("FILE FORMAT NOT RECOGNIZED")
+        self.write_to_s3()
+
+    def nasa_client(self, endpoint):
+        """Client component for API call to NASA endpoint.
+
+        Args:
+            endpoint (str): Parameterized url for API call.
+        Raises:
+            Timeout: If connection not made in 5s and/or data not retrieved in 15s.
+            HTTPError & Exception: Self-explanatory
+        Notes:
+            See Cloudwatch logs for debugging.
+        """
+        try:
+            response = requests.get(endpoint, timeout=(5, 15))
+        except Timeout as timeout:
+            print(f"NASA GET request timed out: {timeout}")
+        except HTTPError as http_err:
+            print(f"HTTP error occurred: {http_err}")
+        except Exception as err:
+            print(f'Other error occurred: {err}')
+        else:
+            return json.loads(response.content)
+
+    def process_asteroids(self, payload):
+        """Process old, and create new, data object with content from response.
+
+        Args:
+            payload (b'str'): Binary string of asteroid data to be processed.
+        """
+        near_earth_objects = payload["near_earth_objects"][f"{self.today}"]
+        asteroids = []
+        for neo in near_earth_objects:
+            asteroid_object = {
+                "id" : neo['id'],
+                "name" : neo['name'],
+                "hazard_potential" : neo['is_potentially_hazardous_asteroid'],
+                "est_diameter_min_ft": neo['estimated_diameter']['feet']['estimated_diameter_min'],
+                "est_diameter_max_ft": neo['estimated_diameter']['feet']['estimated_diameter_max'],
+                "miss_distance_miles": [item['miss_distance']['miles'] for item in neo['close_approach_data']],
+                "close_approach_exact_time": [item['close_approach_date_full'] for item in neo['close_approach_data']]
+            }
+            asteroids.append(asteroid_object)
+
+        return asteroids
+
+    def asteroids_json(self, payload):
+        """Creates json object from payload content then writes to .json file.
+
+        Args:
+            payload (b'str'): Binary string of asteroid data to be processed.
+        """
+        json_file = open(f"/tmp/asteroids_{self.today}.json",'w')
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *json\_file*. The resource is allocated by call *builtins.open*. The resource is referenced at lines 103, 104. The resource is closed at line 104. There are other execution paths that do not contain closure statements, for example, when *TextIOWrapper.write()* throws an exception. To prevent this resource leak, close *json\_file* in a try-finally block or declare it using a `with` statement.

**More info**
[View details about the `with` statement in the Python developer&#39;s guide](https://www.python.org/dev/peps/pep-0343/) (external link).